OCast/src/main/java/org/ocast/core/ReferenceDevice.kt

```diff
@@ -72,44 +88,75 @@ open class ReferenceDevice(upnpDevice: UpnpDevice) : Device(upnpDevice), WebSock
-            State.DISCONNECTED -> {
-                connect({
-                    startDialApp(onSuccess, onError)
-                }, {
-                    onError.wrapRun(it)
-                })
+        applicationName?.ifNotNull { applicationName ->
+            when (state) {
+                State.CONNECTING -> onError.wrapRun(OCastError("Device is connecting, start cannot be processed yet"))
+                State.CONNECTED -> startApplication(applicationName, onSuccess, onError)
+                State.DISCONNECTING -> onError.wrapRun(OCastError("Device is connecting, start cannot be processed."))
+                State.DISCONNECTED -> onError.wrapRun(OCastError("Device is disconnected, start cannot be processed."))
             }
+        }.orElse {
+            onError.wrapRun(OCastError("Property applicationName is not defined"))
         }
     }
 
-    private fun startDialApp(onSuccess: Runnable, onError: Consumer<OCastError>) {
-        // TODO a faire et set isApplicationRunning sur success
-
-        isApplicationRunning = true
+    private fun startApplication(name: String, onSuccess: Runnable, onError: Consumer<OCastError>) {
+        dialClient.getApplication(name) { result ->
+            result
+                .onFailure { throwable ->
+                    onError.wrapRun(OCastError("Failed to start $name, there was an error with the DIAL application information request", throwable))
+                }
+                .onSuccess { application ->
+                    isApplicationRunning.set(application.state == DialApplication.State.Running)
+                    if (isApplicationRunning.get()) {
+                        onSuccess.wrapRun()
+                    } else {
+                        dialClient.startApplication(name) { result ->
+                            result
+                                .onFailure { throwable ->
+                                    onError.wrapRun(OCastError("Failed to start $name, there was an error with the DIAL request", throwable))
+                                }
+                                .onSuccess {
+                                    applicationSemaphore = Semaphore(0)
+                                    if (applicationSemaphore?.tryAcquire(60, TimeUnit.SECONDS) == true &&
+                                        // Semaphore is released when state or application name changes
+                                        // In these cases onError must be called
+                                        applicationName == name &&
+                                        state == State.CONNECTED) {
+                                        onSuccess.wrapRun()
+                                    } else {
+                                        onError.wrapRun(OCastError("Failed to start $name, the WebAppConnectedStatus event was not received"))
+                                    }
+                                }
```

Comment:
remove the bad indent, use instead: :

result.onFailure {
    .....
}
result. onSuccess {
    .......
}

Reply:
Fixed